tools/@aws-cdk/eslint-plugin/lib/rules/no-invalid-path.ts

```diff
@@ -0,0 +1,79 @@
+import { Rule } from 'eslint';
+import { isProdFile } from '../private/is-prod-file';
+import * as path from 'path';
+import * as fs from 'fs';
+
+function isPathJoinFuncCall(node: any): boolean {
+  return (
+    node.callee &&
+    node.callee.property &&
+    node.callee.property.name === 'join' &&
+    node.arguments.length > 1 &&
+    node.arguments[0].name &&
+    node.arguments[0].name === '__dirname'
+  )
+}
+
+export function create(context: Rule.RuleContext): Rule.NodeListener {
+  return {
+    CallExpression(node: any) {
+      if (!isProdFile(context.getFilename())) {
+        return;
+      }
+
+      if (isPathJoinFuncCall(node)) {
+        // Confirm path does not have unnecessary '..'
```

Comment:
First: is this check against oscillating worth the squeeze? (Or is this just to make the subsequent checks easier? If so, I can dig that. Might want to write that down.)

Second: can this not be simplified in the following way:

- `..` may only occur in the path prefix.
- So any `..` whose index is after the first non-`..` is incorrect:

```ts
const components = confirmFirstArgumentIsDirnameAndRemove(node.arguments);

const firstDownDir = components.findIndex((p) => p !== '..');
if (components.some((p, i) => p === '..' && i > firstDownDir) {
  // ERROR: Oscillating paths
  return;
}
```

Then we can observe that `package.json` may only exist in the very last directory that we ended up with by doing `..`: if it occurs anywhere earlier that's an error:

```ts
// Note the - 1 to exclude the last directory from the check
// Exclude the case where there are no '..' at all in the path -- those are never invalid
if (firstDownDir > 0) {
  for (let i = 0; i < firstDownDir - 1; i++) {
    const pjFile = path.join(...[path.dirname(currentFile), ...components.slice(0, i), 'package.json']);
    if (existsSync(pjFile)) {
      // ERROR: this path will end up going out of the package.json directory
      return;
    }
  }
}
```

Reply:
Oh you should also add a check that `/` doesn't occur anywhere inside any of the components, otherwise our checks are going to fail:

```
path.join(__dirname, '../../../..', 'banana')
```

Is bad style but works, and anything that accidentally works will occur in our code base 😅. We need to guard against it.